dashboard/app/controllers/aichat_controller.rb

```diff
@@ -15,24 +15,46 @@ def chat_completion
     end
 
     # Check for PII / Profanity
-    # Copied from ai_tutor_interactions_controller.rb - not sure if filtering is working.
     locale = params[:locale] || "en"
-    new_message_text = params[:newMessage]
-    # Check only the newest message from the user for inappropriate content.
-    filter_result = ShareFiltering.find_failure(new_message_text, locale) if new_message_text
-    # If the content is inappropriate, we skip sending to endpoint and instead hardcode a warning response on the front-end.
-    return render(status: :ok, json: {status: filter_result.type, flagged_content: filter_result.content}) if filter_result
-
-    input_json = AichatSagemakerHelper.format_inputs_for_sagemaker_request(params[:aichatModelCustomizations], params[:storedMessages], params[:newMessage])
-    sagemaker_response = AichatSagemakerHelper.request_sagemaker_chat_completion(input_json)
+    filter_result = ShareFiltering.find_failure(params[:newMessage], locale)
+    if filter_result&.type == 'profanity'
+      new_messages = [
+        {
+          role: 'user',
+          content: params[:newMessage],
+          status: SharedConstants::AI_INTERACTION_STATUS[:PROFANITY_VIOLATION]
+        }
+      ]
+      session_id = log_chat_session(new_messages)
+      return render(
+        status: :ok,
+        json: {
+          status: filter_result.type,
+          flagged_content: filter_result.content,
+          session_id: session_id
+        }
+      )
+    end
+
+    # Use to_unsafe_h here to allow testing this function.
+    # Safe params are primarily targeted at preventing "mass assignment vulnerability"
+    # which isn't relevant here.
+    input = AichatSagemakerHelper.format_inputs_for_sagemaker_request(
+      params.to_unsafe_h[:aichatModelCustomizations],
+      params.to_unsafe_h[:storedMessages].filter {|message| message[:status] == SharedConstants::AI_INTERACTION_STATUS[:OK]},
+      params.to_unsafe_h[:newMessage]
+    )
+    sagemaker_response = AichatSagemakerHelper.request_sagemaker_chat_completion(input)
     latest_assistant_response = AichatSagemakerHelper.get_sagemaker_assistant_response(sagemaker_response)
-    assistant_message = {
-      role: "assistant",
-      content: latest_assistant_response,
-    }
 
-    session_id = log_chat_session(assistant_message)
-    return render(status: :ok, json: assistant_message.merge({sessionId: session_id}).to_json)
+    assistant_message = {role: "assistant", content: latest_assistant_response, status: SharedConstants::AI_INTERACTION_STATUS[:OK]}
+    new_messages = [
+      {role: 'user', content: params[:newMessage], status: SharedConstants::AI_INTERACTION_STATUS[:OK]},
+      assistant_message,
+    ]
+    session_id = log_chat_session(new_messages)
+
+    render(status: :ok, json: assistant_message.merge({session_id: session_id}).to_json)
```

Comment:
super minor: should we keep `sessionId` camel case for the front end?

Reply:
haha I switched to the opposite :) It felt consistent with the fact we don't "tee up" the back end by converting all params we send to Rails into snake case? Relatedly, I was thinking it'd be nice to do something like what AI Tutor is doing in their controller:

https://github.com/code-dot-org/code-dot-org/blob/f280b95523fb6a3aabbe624250146256a9480cef/dashboard/app/controllers/ai_tutor_interactions_controller.rb#L18-L26

Reply:
Ahh gotcha, yep not a huge deal either way and I do like what AI Tutor is doing in the snippet you linked, so maybe we can take a pass at cleaning this up later